### PR TITLE
Add weapon strafing support and implement it on RA Yak

### LINKED ISSUE #2642: Add strafing logic
The original RA had logic in which the yak would instead of dumping all of its ammo in one spot, it would dump it in a line. We should add this feature into openRA, with the following features:

If targeting a unit, the strafing unit will first dump ammo on the tile the unit is on, and continue on.
If targeting a building, the strafing unit will first dump ammo on the occupied tile that the strafing unit is closest to.
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/100103-add-strafing-logic?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #2730: Bursts reload when idle
If a unit has burst but they have not fully spent all their bursts by firing, the next time they fire they will only have the remaining bursts left to fire with. This can be disadvantage if your have a lot of bursts, but only have one shot left. After firing, you then have to wait a long time for the unit to reload again.

Proposal: Say you have finished firing but still have unfired bursts remaining. You wait for your ROF time to pass, and then your unit begins to reload, one shot at a time. Example: Say you have Burst:4  and ROF:100. You fire 3 shots, and have one left. Your unit then pauses for 100 ticks. At that point, one burst is reloaded every X ticks, where X is the BurstDelay value (default=5). So the total reload time would be 100 + 5 + 5 + 5 = 115.
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/115665-bursts-reload-when-idle?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by reaperrr**
This has been overdue in my opinion.

Modifying the targeted offset per burst is by far the easiest approach compared to modifying Attack traits or projectiles.

Closes #2642.
Fixes #2730.

### CODE REVIEW

**Comment by reaperrr** [comment]
Pinging @SoScared and @OpenRA/engine-hackers for feedback on whether we might want this for the playtest (or not at all, at least wrt using it in RA).

**Comment by reaperrr** [comment]
Moving it to playtest for now, as it would be best to include it together with the hit-shape changes so balance can be revamped in one go.

**Comment by pchote** [comment]
Needs a rebase.

**Review by pchote — COMMENTED** [review]
Two things I noticed while testing:
* The yak now (usually) empties its entire ammo pool in the first past, instead of spreading it over two passes.  We should keep it at two passes before reloading.
* Sometimes it only fires one blip in the first pass.  Probably related to #13220 and so out of scope of this PR, but still annoying.

Otherwise, the code implementation looks good.

It would be good if you could remove the hardcoded strafing logic from `AttackBomber` and make it use this instead.  It should be possible to remove the custom `Guns` and `Bombs` properties and use `Armaments` from `AttackBase`.

**Comment by reaperrr** [comment]
Rebased and updated.

I think I fixed `Sometimes it only fires one blip in the first pass. Probably related to #13220 and so out of scope of this PR, but still annoying.`, though I didn't fully took the approach suggested in #13220 as completely replacing the FireDelay/ReloadDelay dance would require covering some edge cases and I didn't want to risk more regressions than necessary.

> It would be good if you could remove the hardcoded strafing logic from AttackBomber and make it use this instead. It should be possible to remove the custom Guns and Bombs properties and use Armaments from AttackBase.

I'll see how far I get on this.

**Comment by reaperrr** [comment]
Updated, ripped out the hardcoded `AttackBomber` strafing logic.

File: OpenRA.Mods.Common/Traits/Armament.cs [hunk 1]
```diff
@@ -221,10 +225,32 @@ public virtual Barrel CheckFire(Actor self, IFacing facing, Target target)
...
 			Func<WPos> muzzlePosition = () => self.CenterPosition + MuzzleOffset(self, barrel);
 			var legacyFacing = MuzzleOrientation(self, barrel).Yaw.Angle / 4;
 
+			var passiveTarget = Weapon.TargetActorCenter ? target.CenterPosition : target.Positions.PositionClosestTo(muzzlePosition());
+			var initialOffset = Weapon.FirstBurstTargetOffset;
+			if (initialOffset != WVec.Zero)
+			{
+				// We want this to match Armament.LocalOffset, so we need to convert it to forward, right, up
+				initialOffset = new WVec(initialOffset.Y, -initialOffset.X, initialOffset.Z);
+				passiveTarget += initialOffset.Rotate(WRot.FromFacing(legacyFacing));
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Armament.cs]: `legacyFacing` is the facing of the barrel, which is not necessarily the same as the facing of the actor.  This makes sense I guess, but might not be what you expect.

File: mods/cnc/weapons/other.yaml [hunk 2]
```diff
@@ -65,7 +65,7 @@ Chemspray:
 Napalm:
 	Inherits: ^FlameWeapon
 	ReloadDelay: 4
-	Range: 2c0
+	Range: 5c0
```
Comment (pchote) [inline mods/cnc/weapons/other.yaml]: What is this for?  Doesn't seem related to this PR.

File: mods/ra/weapons/smallcaliber.yaml [hunk 3]
```diff
@@ -201,16 +201,24 @@ ChainGun:
 		Versus:
 			None: 120
 
-ChainGun.Yak:
+ChainGun.Yak.Right:
 	Inherits: ^HeavyMG
-	ReloadDelay: 3
-	Range: 5c0
-	MinRange: 3c0
+	Burst: 5
```
Comment (pchote) [inline mods/ra/weapons/smallcaliber.yaml]: This won't evenly divide into 18, but nothing will so we can leave this for a followup balance patch to hopefully iron out.

**Review by pchote — COMMENTED** [review]
Looks good.  Just one blocking question / probably change request, and a couple of info points:

File: mods/cnc/weapons/other.yaml (hunk 2 above)
Comment (reaperrr) [inline mods/cnc/weapons/other.yaml]: I thought this was necessary to make the aircraft start dropping bombs early enough, but I admit I don't remember testing it.

File: mods/ra/weapons/smallcaliber.yaml (hunk 3 above)
Comment (reaperrr) [inline mods/ra/weapons/smallcaliber.yaml]: In my opinion the most sensible fix is (probably) to increase ammo to 20, I didn't want to make that call on my own.

File: OpenRA.Mods.Common/Traits/Armament.cs (hunk 1 above)
Comment (reaperrr) [inline OpenRA.Mods.Common/Traits/Armament.cs]: If you're fine with it, I'd actually like to leave it like this. If I or any other modder ever needs a different behavior, we can still fix it (or make it configurable).

**Comment by reaperrr** [comment]
Updated, TD airstrike now behaves pretty much like on bleed again.

Additionally, I added balance commits for TD airstrike and RA parabomb strike, because on bleed they are so obviously over-powered vs. larger building shapes now, that not making changes before the playtest would probably just result in testers focusing too much on those airstrikes.
These changes are obviously not meant to be a final solution, they just make sure that airstrikes don't 1HKO cyards.

File: OpenRA.Mods.Common/Traits/Armament.cs [hunk 4]
```diff
@@ -169,6 +171,8 @@ protected virtual void Tick(Actor self)
 			if (IsTraitDisabled)
 				return;
 
+			++ticksSinceLastShot;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Armament.cs]: Style nit: Try to avoid incrementing an integer without bound because it could in principle overflow and become negative.  Thats not likely in practice to happen here, but it is still bad style.  You can put this behind an `if (ticksSinceLastShot < Weapon.ReloadDelay)`.

File: OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs [hunk 5]
```diff
@@ -749,6 +749,22 @@ internal static void UpgradeActorRules(ModData modData, int engineVersion, ref L
...
+				if (engineVersion < 20170624)
+				{
+					if (node.Key == "AttackBomber")
+					{
+						var gunsOrBombs = node.Value.Nodes.FirstOrDefault(n => n.Key == "Guns" || n.Key == "Bombs");
+						if (gunsOrBombs != null)
+						{
+							Console.WriteLine("Hardcoded Guns and Bombs logic has been removed from AttackBomber.");
+							Console.WriteLine("Bombs should work like usual, for gun strafing use the new Weapon TargetOffset modifiers.");
+							Console.WriteLine("Look at the C&C1 mod's A10 for an example.");
```
Comment (pchote) [inline OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs]: Style nit: C&C1 &rarr; TD.

**Review by pchote — APPROVED** [review]
Looks good :+1:. Just a couple of minor style nits:

**Comment by reaperrr** [comment]
Fixed.

**Comment by FrameLimiter** [comment]
Just gave this a quick test and noticed the damage vs infantry is far too low.  It takes two full strafes (full ammo payload) to kill just 1 Flame thrower (40hp).  Also killing artillery with yaks is far less effective now; before you could kill four lined up artillery in one pass, but now you can only kill 1 of them.

I think both the strafe damage & the ability to queue up multiple targets could use some tweaking.

**Comment by reaperrr** [comment]
@FrameLimiter I added some preliminary balance changes, please re-test. Note that they don't need to be perfect for the playtest, just 'good enough', though.

**Comment by abcdefg30** [comment]
Needs a rebase again.

**Comment by abcdefg30** [comment]
I'm really not sure if we want this enabled on the  yak. You can't easily kill two arties in one run anymore.
That aside 👍 / ✅ .

**Comment by reaperrr** [comment]
Rebased.

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/0a4ac6bb1981753dbbf3e017ae6a51ded1a29061...aad16a0588898957ab59b987c5b2946e7b59e587)

**Comment by SoScared** [comment]
Late to the party but my thoughts are as follows:

I've tried to keep an open mind but I'm not sure the Yak change is worth pandering towards the original game in this case. The repurposing is blind and I don't see a relevant problem with #2642 #2730. The RA mod meta has developed heavily around the current Yak and I'm not sure if it's wise to bring this aboard the official playtest if we want players to dive into the HitShape/UnitStance seriously.
